cronsnoop: Add plugin which is able to snoop removal/addition of cron…

lorddoskias opened the pull request:
… entries

lorddoskias committed:
cronsnoop: Add plugin which is able to snoop removal/addition of cron entries

jeffmahoney commented on vql/linux/cronsnoop/plugin.go:
```diff
@@ -0,0 +1,88 @@
+		// Make the chan large enough so that cron snooper doesn't block, waiting 
+		// for the channel too be consumed
+		eventChan := make(chan CronEvent, 1000)
+		snooper, err := NewCronSnooperWithChan(arg.SpoolPath, arg.SystemPaths, eventChan)
```
You might look at using the lfreequeue implementation I imported for use in audit. I ended up using it precisely for this reason.

lorddoskias replied:
Oh the joys of random packages on the internet... I pulled in the lfreequeue and it turned out its watchIterator actually cannot be created, since the type is unexported (watchIterator vs WatchIterator, I guess the original author never tested this). Anyway, fixed that (which means if we are to use it, we'll likely have to pull it as a 3rd party dependency and fix that oversight in the local copy) and indeed the lfreequeue can be used however, it leads to a slight inconvenience that the caller should be doing type assertions to get a CronEvent vs the generic interface {} with which lfreequeue has been defined. Of course this could be fixed by running a go routine that does the actual dequeue + type assertion and pumps stuff to properly typed channel and in turn this channel is the one returned to the user but at this point I'm starting to debate the utility of all this. One of the tests adjusted leads to the following: 

```
    // returns the WatchIterator's Iter()'s channel
    c := s.GetEventChan()                                                       
                                                                                                                                           
    v := <-c                       // reads a  generic interface {} typec                                         
    v1 := v.(CronEvent)     // do the typeassertion (casting)
                                                 
     // assertCronEvent takes a CronEvent as the first argument                                                                           
    assertCronEvent(v1, cmd[0], fName, fName, Added, t)                         
    v = <-c                                                                     
    v1 = v.(CronEvent)                                                          
    assertCronEvent(v1, cmd[1], fName, fName, Added, t)       
```

jeffmahoney commented on vql/linux/cronsnoop/plugin.go:
```diff
@@ -0,0 +1,88 @@
+
+	output_chan := make(chan vfilter.Row)
+	arg := &CronsnoopArgs{}
+    err := arg_parser.ExtractArgsWithContext(ctx, scope, args, arg)
```
Weird indentation

jeffmahoney commented on vql/linux/cronsnoop/cron-lib.go:
```diff
@@ -0,0 +1,395 @@
+		filebase := filepath.Base(event.Name)
+		filebase_c := C.CString(filebase)
+		defer C.free(unsafe.Pointer(filebase_c))
+		cpw := C.getpwnam(filebase_c)
```
https://pkg.go.dev/os/user might make more sense here

jeffmahoney commented on vql/linux/cronsnoop/cron-lib.go:
```diff
@@ -0,0 +1,395 @@
+
+	// user got all of its entries deleted so just delete the user altogether
+	if len(existing) == 0 {
+		//TODO: Emit an event that all of user's entries are deleted
```
Is this a TODO you want to do before merge?

lorddoskias replied:
So looking at the todo and the surrounding code all the of user's commands would have already been emitted as being deleted due to :

```
   // emit events                                                              
    for _, deleted_cmd := range deleted {                                       
        snooper.emit_event(deleted_cmd, user, path, Deleted)                    
    }  
```

So even if I don't do anything for the todo we are not missing any information w.r.t what entries got deleted, however we won't get anything explicitly saying that all commands got deleted. Imo this is not a big deal. What do you say ?

jeffmahoney commented on vql/linux/cronsnoop/cron-lib.go:
```diff
@@ -0,0 +1,395 @@
+
+		// TODO: Probably could optimise how we get the command instead of doing
+		// split + join
+		if special_cron_string(split_line) {
```
Maybe, but this isn't really performance-critical code, is it?

lorddoskias replied:
On a second pass you are right.

jeffmahoney reviewed (changes requested):
Minor cleanups. Use os.user instead of calling C.getpwnam directly.

lorddoskias committed:
cronsnoop: rewrite is_user_cron_event using go-code

Instead of using the cgo interface to glibc functions use the native
go implementation.

lorddoskias committed:
misc cleanups

jeffmahoney committed:
SUSE.Linux.Events.Cron: fix stray double quote
